packages/light.js/src/rpc/utils/createRpc.ts

```diff
@@ -4,70 +4,33 @@
 // SPDX-License-Identifier: MIT
 
 import { isFunction } from '@parity/api/lib/util/types';
+import * as memoizee from 'memoizee';
 import { merge, ReplaySubject, Observable, OperatorFunction } from 'rxjs';
 import { multicast, refCount } from 'rxjs/operators';
-import * as prune from 'json-prune';
 
 import { createApiFromProvider, getApi } from '../../api';
-import {
-  distinctValues,
-  withoutLoading as withoutLoadingOperator
-} from '../../utils/operators';
-import { Metadata, RpcObservable, RpcObservableOptions } from '../../types';
-
-interface RpcObservableWithoutMetadata<_, Out> {
-  (...args: any[]): Observable<Out>;
-}
-
-/**
- * Mixins that are added into an RpcObservable.
- *
- * @ignore
- */
-const frequencyMixins = {
-  /**
-   * Change the frequency of a RPC Observable.
-   *
-   * @param frequency - An array of frequency Observables.
-   * @example
-   * balanceOf$.setFrequency([onEverySecond$, onStartup$]); // Will fetch
-   * balance once on startup, and then every second.
-   */
-  setFrequency(frequency: Observable<any>[]) {
-    // TODO Check that frequency is well-formed
-
-    this.metadata.frequency = frequency;
-
-    // If necessary, we clear the memoize cache
-    if (typeof this.clear === 'function') {
-      this.clear();
-    }
-  }
-};
+import { distinctValues } from '../../utils/operators';
+import { Metadata, RpcObservableOptions } from '../../types';
 
 /**
  * Add metadata to an RpcObservable, and transform it into a ReplaySubject(1).
  * It's a currified function.
+ * Pure function version of {@link createRpc}.
  *
  * @ignore
  * @param metadata - The metadata to add.
  * @example
  * createRpc(metadata)(options) returns a RpcObservable.
  * createRpc(metadata)(options)(someArgs) returns an Observable.
  */
-const createRpc = <Source, Out>(metadata: Metadata<Source, Out>) => (
-  options: RpcObservableOptions = {}
-) => {
-  const { provider, withoutLoading } = options;
-  const api = provider ? createApiFromProvider(provider) : getApi();
-  // rpc$ will hold the RpcObservable minus its metadata
-  const rpc$: RpcObservableWithoutMetadata<Source, Out> = (...args: any[]) => {
+const createRpcWithApi = memoizee(
+  <Source, Out>(metadata: Metadata<Source, Out>, api: any, ...args: any[]) => {
     // The source Observable can either be another RpcObservable (in the
     // `dependsOn` field), or anObservable built by merging all the
     // FrequencyObservables
     const source$ = metadata.dependsOn
-      ? metadata.dependsOn(...args, options)
-      : merge(...metadata.frequency.map(f => f(provider)));
+      ? metadata.dependsOn(...args, { provider: api.provider })
+      : merge(...metadata.frequency.map(f => f({ provider: api.provider })));
```

Comment:
isn't the fact that we're passing the provider around instead of the api going to cause that we will have a different websocket connection for each different rpc endpoint? (not sure where the ws connection is initialized/stored in `@parity/api` Api/Provider)

Reply:
not a problem: createApiFromProvider is memoized